### ISSUE TITLE
Audio element stops/pause playing even if it is flag as data-turbolinks-permanent

### ISSUE THREAD

**Issue by rgagnon** [turbolinks/turbolinks#157]
Hi,

I'm trying to make turbolinks 5 to work with an html5 audio element.  I put this element within a div flagged as data-turbolinks-permanent.  Everything works well, the div has not been refresh but the audio just pause.  I can hit the play and it continues where it was.

It was ok with a previous version of turbolinks.

Any idea?

**Comment by beastafk** [comment]
Hi,

I don't know about previous versions, but I had the same problem with video elements and webRTC live communication. I managed to "solve" it by manually calling $meMediaElement.play(); on every `turbolinks:load` event, something like that

```
document.addEventListener("turbolinks:load", function() {
    $('.js-local-stream')[0].play();
    $('.js-remote-stream')[0].play();
});
```

It's a stupid fix, but I haven't found anything better for now and for me it works. There is a problem though with normal `<video>`  elements. While using the solution from above for playing the video works, the audio is started each time from the beginning, which causes a strange case, where you hear your audio form the beginning on every page visit. Something in the turbolinks is making the media elements stop on load.

**Comment by rgagnon** [comment]
Thanks, 

I did  the same thing on my side though.  My issue with this is if the user pause the audio before changing page, the audio starts playing.

Glad I'm not alone with this issue.

**Comment by christophwolff** [comment]
You can solve this by creating a plain audio object without a <audio> Tag

`const audio = new Audio('path/to/mp3')
audio.play();
`

Then changing pages works. But you need to build an own audio element with play pause, volume etc..

**Comment by nateberkopec** [comment]
Similar to #177.

**Comment by Hainminsoe** [comment]
same here ..

**Comment by domchristie** [comment]
Turbolinks clones `data-turbolinks-permanent` elements before rendering a visit, then replaces elements in the new `<body>` with the cloned ones. It looks like the "playing" state is excluded from the clone 😞 

Good news is that uninterrupted audio/video play back across pages can be achieved by tracking the playing state ourselves. I've not been able to discover how media elements are cloned, but (luckily) it seems like the current time _is_ cloned, so we can just call `play` (if applicable) once the visit has rendered 🎉 

```js
;(function () {
  var forEach = Array.prototype.forEach
  var selector = [
    '[data-turbolinks-permanent] audio,',
    'audio[data-turbolinks-permanent]',
    '[data-turbolinks-permanent] video',
    'video[data-turbolinks-permanent]'
  ].join(',')

  addEventListener('turbolinks:load', function () {
    var mediaElements = document.querySelectorAll(selector)
    forEach.call(mediaElements, uninterrupt)
  })

  // Bind to turbolinks:render to handle previews
  addEventListener('turbolinks:render', function () {
    var mediaElements = document.querySelectorAll(selector)
    forEach.call(mediaElements, function (media) {
      if (media.isPlaying) media.play()
    })
  })

  function uninterrupt (media) {
    if (!media.uninterrupted) {
      media.addEventListener('play', function () { media.isPlaying = true })
      media.addEventListener('pause', function () { media.isPlaying = false })
      media.uninterrupted = true
    }
  }
})()
```

This will work for `data-turbolinks-permanent` media elements, and media elements contained within a `data-turbolinks-permanent`container. It seemed to operate seemlessly on Safari and Firefox macOS, and with minor clicks/jumping when switching between pages on Chrome. Hope that helps!

**Comment by mcalia** [comment]
Hi all,
I'm using turbolinks by a wordpress plugin (https://it.wordpress.org/plugins/turbolinks/).

I faced the same issue (music stops playing). But it can be resolved by specifying the **ID** attribute of the "data-turbolinks-permanent" container.

This is my case:
I implemented a very simple test audio player (just for testing turbolinks) by quickly adding this in my theme's header.php file:
		
```
<!-- NOT WORKING example -->
<div data-turbolinks-permanent>
	<audio controls src="mysong.mp3">
	  Your browser does not support the audio element.
	</audio>
</div>
```	
This way, any time I click on a link, the player stops playing.

if I change the code this way, it works perfectly and the audio keeps playing when moving from a page to another:
```
<!-- WORKING FINE example -->
<div id="my-div" data-turbolinks-permanent>
	<audio controls src="mysong.mp3">
	  Your browser does not support the audio element.
	</audio>
</div>
```	

It seems that the ID attribute is necessary on the turbolinks' permanent containers.

Hope this helps.
